lib/virtio_mmio/virtio_mmio_drv.c

```diff
@@ -0,0 +1,393 @@
+static void _virtio_mmio_set_features(struct virtio_device *vdev,
+				      uint32_t features, int idx)
+{
+	uint32_t hfeatures;
+
+	/* Writing selection register VIRTIO_MMIO_DEVICE_FEATURES_SEL. In pure AMP
+	 * mode this needs to be followed by a synchronization w/ the device
+	 * before reading VIRTIO_MMIO_DEVICE_FEATURES
+	 */
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_DEVICE_FEATURES_SEL, idx);
+	hfeatures = virtio_mmio_read32(vdev, VIRTIO_MMIO_DEVICE_FEATURES);
+	features &= hfeatures;
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_DRIVER_FEATURES, features);
+	vdev->features = features;
+}
+
+static void virtio_mmio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	_virtio_mmio_set_features(vdev, features, 0);
+}
+
+static void virtio_mmio_reset_device(struct virtio_device *vdev)
+{
+	virtio_mmio_set_status(vdev, 0);
+}
+
+static void virtio_mmio_notify(struct virtqueue *vq)
+{
+	/* VIRTIO_F_NOTIFICATION_DATA is not supported for now */
+	virtio_mmio_write32(vq->vq_dev, VIRTIO_MMIO_QUEUE_NOTIFY, vq->vq_queue_index);
+}
+
+const struct virtio_dispatch virtio_mmio_dispatch = {
+	.create_virtqueues = virtio_mmio_create_virtqueues,
+	.get_status = virtio_mmio_get_status,
+	.set_status = virtio_mmio_set_status,
+	.get_features = virtio_mmio_get_features,
+	.set_features = virtio_mmio_set_features,
+	.read_config = virtio_mmio_read_config,
+	.write_config = virtio_mmio_write_config,
+	.reset_device = virtio_mmio_reset_device,
+	.notify = virtio_mmio_notify,
+};
+
+static int virtio_mmio_get_metal_io(struct virtio_device *vdev, uintptr_t virt_mem_ptr,
+				    uintptr_t cfg_mem_ptr)
+{
+	struct metal_device *device;
+	int32_t err;
+	int device_idx = 0;
+	struct virtio_mmio_device *vmdev = metal_container_of(vdev,
+							      struct virtio_mmio_device, vdev);
+
+	/* Setup shared memory device */
+	vmdev->shm_device.regions[0].physmap = (metal_phys_addr_t *)&vmdev->shm_mem.base;
+	vmdev->shm_device.regions[0].virt = (void *)virt_mem_ptr;
+	vmdev->shm_device.regions[0].size = vmdev->shm_mem.size;
+
+	VIRTIO_ASSERT((METAL_MAX_DEVICE_REGIONS > 1),
+		      "METAL_MAX_DEVICE_REGIONS must be greater that 1");
+
+	vmdev->shm_device.regions[1].physmap = (metal_phys_addr_t *)&vmdev->cfg_mem.base;
+	vmdev->shm_device.regions[1].virt = (void *)cfg_mem_ptr;
+	vmdev->shm_device.regions[1].size = vmdev->cfg_mem.size;
+
+	err = metal_register_generic_device(&vmdev->shm_device);
+	if (err) {
+		metal_log(METAL_LOG_ERROR, "Couldn't register shared memory device: %d\n", err);
+		return err;
+	}
+
+	err = metal_device_open("generic", vmdev->shm_device.name, &device);
+	if (err) {
+		metal_log(METAL_LOG_ERROR, "metal_device_open failed: %d", err);
+		return err;
+	}
+
+	vmdev->shm_io = metal_device_io_region(device, device_idx);
+	if (!vmdev->shm_io) {
+		metal_log(METAL_LOG_ERROR, "metal_device_io_region failed to get region %d",
+			  device_idx);
+		return err;
+	}
+
+	vmdev->cfg_io = metal_device_io_region(device, 1);
```

Comment:
We can use device_idx variable here, instead of 1. 
Or we can remove device_idx all together and use only hardcode index.

Reply:
Removed device_idx.